incubator/orm/natural_key.go

```diff
@@ -3,101 +3,97 @@ package orm
 import (
 	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/errors"
 )
 
 var _ TableBuilder = &naturalKeyTableBuilder{}
 
-type naturalKeyer func(val interface{}) []byte // todo: note: in the api design this does not return an error unlike other indexer functions do
-
-func NewNaturalKeyTableBuilder(prefixData, prefixSeq, prefixIndex byte, key sdk.StoreKey, cdc *codec.Codec, model interface{}, getPrimaryKey naturalKeyer) *naturalKeyTableBuilder {
+func NewNaturalKeyTableBuilder(prefixData, prefixSeq, prefixIndex byte, key sdk.StoreKey, cdc *codec.Codec, model NaturalKeyed) *naturalKeyTableBuilder {
```

Comment:
I definitely prefer having the `NaturalKeyed` method on the model, rather than some random function. It is always computed from the model itself, so it makes most sense (and cleaner code) to just place it there